### PR TITLE
Remove special handling of obsolete `impl Trait for ..` syntax

### PR DESCRIPTION
**Pull request by nnethercote**
The ancient (pre-1.0) RFC 19 suggested using `impl Trait for ..` syntax for default traits. That was later changed to `auto trait Trait {}` syntax. The parser has special treatment for the `..` syntax, suggesting the `auto` syntax.

Given that default traits have not be stabilized and the `..` syntax is so old, the special case seems unnecessary, and it gets in the way of adding `ErrorGuaranteed` to `TyKind::Err`. This commit removes it and the tests.

r? @oli-obk

### CODE REVIEW

**Comment by nnethercote** [comment]
Best reviewed one commit at a time.

File: compiler/rustc_parse/src/parser/item.rs [hunk 1]
```diff
@@ -588,15 +588,8 @@ impl<'a> Parser<'a> {
         let has_for = self.eat_keyword(kw::For);
         let missing_for_span = self.prev_token.span.between(self.token.span);
 
-        let ty_second = if self.token == token::DotDot {
-            // We need to report this error after `cfg` expansion for compatibility reasons
```
Comment (Nilstrieb) [inline compiler/rustc_parse/src/parser/item.rs:592]: This suggests that we intend the following (historically wrongly gated 😭) to pass:
```rust
#[cfg(dont)]
impl Trait for .. {}
```
given that this should have emitted a future compat warning (from a quick test it doesn't seem to do that though????, I might be doing it wrong though)

but this is a breaking change (a probably justified one, but one)

maybe this should just use some dummy node in the meantime to not block this

**Comment by petrochenkov** [comment]
This is related to https://github.com/rust-lang/rust/issues/65860.

**Comment by oli-obk** [comment]
@craterbot check

**Comment by nnethercote** [comment]
I don't want to hold up the `TyKind::Err` changes on the `impl Trait for ..` syntax changes. So I've filed #121109 for just the `TyKind::Err` changes, and this PR can morph into being about the `..` syntax changes.

**Comment by nnethercote** [comment]
The crater run has 77 failures, which boil down to three crates.

### crossbeam

Most of the failures are from this crate, versions 0.2.9, 0.2.10, 0.3.0, which are all over six years old. Current version is 0.8.4.
```
[INFO] [stdout] error: expected type, found `..`
[INFO] [stdout]   --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/crossbeam-0.2.9/src/mem/cache_padded.rs:41:28
[INFO] [stdout]    |
[INFO] [stdout] 41 | unsafe impl ZerosValid for .. {}
[INFO] [stdout]    |                            ^^ expected type
```

```
[INFO] [stdout] error: expected type, found `..`
[INFO] [stdout]   --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/crossbeam-0.2.10/src/mem/cache_padded.rs:49:28
[INFO] [stdout]    |
[INFO] [stdout] 49 | unsafe impl ZerosValid for .. {}
[INFO] [stdout]    |                            ^^ expected type
```

```
[INFO] [stdout] error: expected type, found `..`
[INFO] [stdout]   --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/crossbeam-0.3.0/src/cache_padded.rs:49:28
[INFO] [stdout]    |
[INFO] [stdout] 49 | unsafe impl ZerosValid for .. {}
[INFO] [stdout]    |                            ^^ expected type
```

### dimensioned

We see a few failures with version 0.6.0, which is seven years old. Current version is 0.8.0, which is two years old. The crate has 220,000 downloads for all time.
```
[INFO] [stdout] error: expected type, found `..`
[INFO] [stdout]    --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/dimensioned-0.6.0/src/traits.rs:129:17
[INFO] [stdout]     |
[INFO] [stdout] 129 | impl NotDim for .. {}
[INFO] [stdout]     |                 ^^ expected type
```

### packed

We see a couple of failures with version 0.4.2, which is the latest version, but over seven years old. The crate has 21,000 downloads for all time.
```
[INFO] [stdout] error: expected type, found `..`
[INFO] [stdout]    --> /opt/rustwide/cargo-home/registry/src/index.crates.io-6f17d22bba15001f/packed-0.4.2/src/lib.rs:141:31
[INFO] [stdout]     |
[INFO] [stdout] 141 |     unsafe impl Unaligned for .. { }
[INFO] [stdout]     |                               ^^ expected type
```

**Comment by nnethercote** [comment]
After some [discussion](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/.E2.9C.94.20.60impl.20Trait.20for.20.2E.2E.60.20syntax/near/422719627) with @tmandry I have decided that removing this is more trouble than it is worth, given that it's a pretty tiny chunk of code in the compiler.

**Pull request closed without merging**